### DISCUSSION TITLE
Uosc customization

### DISCUSSION THREAD

**Discussion by ElvioG** [tomasklaen/uosc#572] [Q&A] [answered]
Hello everyone,
I have been using MPV with the UOSC ui for a while now, I have customized some functions of MPV and wanted to know if it was possible to fix or customize some things.

- When I open an audio file (mp3, wav) I would like to have a smaller window size, and change the window background (is black now)

- I would like to add a background in alpha behind the osc, so that more buttons and elements contrast especially if I have a dark scene.

- When I have a very short video or audio that lasts only a few seconds, on the seekbar I display (e.g. for 10 seconds) 01-10, instead I would like to be able to display in detail, such as h:m:s

- Is possible with MPV have custom icons per file type, like Pot Player?

Thank you

**Comment by christoph-heinrich** [comment]
>     * When I open an audio file (mp3, wav) I would like to have a smaller window size, and change the window background (is black now)

You can use [conditional auto profiles](https://mpv.io/manual/master/#conditional-auto-profiles) to change the window size using [geometry](https://mpv.io/manual/master/#conditional-auto-profiles) or [autofit](https://mpv.io/manual/master/#conditional-auto-profiles).
```
[audio-only]
profile-cond=vid == false
autofit=500
profile-restore=copy-equal
```
The black window can be remedied with the [visualizer.lua](https://mpv.io/manual/master/#conditional-auto-profiles) script.

>     * When I have a very short video or audio that lasts only a few seconds, on the seekbar I display (e.g. for 10 seconds) 01-10, instead I would like to be able to display in detail, such as hms

That's currently not configurable. Do we want it to be? @tomasklaen
>     * Is possible with MPV have custom icons per file type, like Pot Player?

I don't think that's possible.

**Reply by ElvioG** [reply]
Thank you!

**Reply by tomasklaen** [reply]
> That's currently not configurable. Do we want it to be?

Well, the only benefit here is that ElvioG likes it that way, but it's more options to maintain and bloat the config file with, and the only thing it accomplishes is make reading the timestamps harder. There's no informational or any kind of value to displaying a bunch of zeroes before actual time value.

@ElvioG the only reason you ask for this is because you're used to it. But objectively the bigger the timestamp is the more work your brain has to do to parse it and figure out what it means. But if it's only 4 numbers, it's immediately obvious it's less than an hour long, 2 numbers means less than a minute. Trimming non-informational parts of timestamps makes parsing the value faster/easier.

**Reply by ElvioG** [reply]
> Well, the only benefit here is that ElvioG likes it that way

I asked because in my journey to find a decent UI on Windows, I found that only UOSC had this thing in the seekbar. Basic MPV shows the h : m : s format even if an mp3 file is 4 seconds long. So I was wondering if it was possible to override this setting via mpv.conf. It is not my whim...

**Reply by ElvioG** [reply]
I understand that maybe you don't like the h : m : s format in short files, but it would be a useful thing to give the option to override it whenever the user feels like customizing their experience to suit their needs and tastes, don't you think?

**Reply by tomasklaen** [reply]
Since it adds an option that needs to be maintained and increases config file bloat, it needs to be justified. It needs to add value.

And you haven't provided any reason why displaying zeroes that don't convey any information adds value, or addressed the fact that doing so makes parsing the smaller timestamps harder/slower. So in this sense, it'd be an option to make things worse.

Your arguments are still a biased "I like it like that", and "others do it too", which don't count.

This is not just my whim. The way it works now was a requested feature that I agreed with, since it makes sense from any angle I look at it. You're just not used to it. Once you do, you'll be annoyed at pointlessly long timestamps, as I am now. It's similar difference as displaying time as `2023-06-01T10:17:37.533Z` instead of `1 second ago`.

**Reply by ElvioG** [reply]
Let me understand something, MPV basic always shows hms format even if the file is short. In UOSC it doesn't, there are only the seconds, so it seems to me that there is your legitimate choice, based on your taste and the way you work, no?
I am involved in audio and video production, and in whatever software I have used in my life that my brain is used to, both visually and in terms of usability, I have always found the hms format, ms, etc., to be the most common. 
Now, considering that it is a feature taken out of the basic MPV program, considering that I know how I work and I don't go to others to tell them how their brains work, considering that I am asking for me, and I am certainly not saying that you have to work and implement an option (also because I never asked to implement an option, but rather if it would be possible to show the hms format), where can I restore the feature that was taken out and changed?
Then if you can't give me this information or if it's not possible just say so, you can't and I'll get over it, but if you allow, my habits, my way of working, my tastes are precisely my own, so if it's possible I ask, if it's not it's okay.
Thank you

**Reply by christoph-heinrich** [reply, marked as answer]
Here is how to change that
```diff
diff --git a/scripts/uosc_shared/lib/utils.lua b/scripts/uosc_shared/lib/utils.lua
index 4eecddc..9847664 100644
--- a/scripts/uosc_shared/lib/utils.lua
+++ b/scripts/uosc_shared/lib/utils.lua
@@ -217,14 +217,6 @@ function format_time(seconds, max_seconds)
 		local formatted = string.format('%.' .. options.time_precision .. 'f', math.abs(seconds) % 1)
 		human = human .. '.' .. string.sub(formatted, 3)
 	end
-	if max_seconds then
-		local trim_length = (max_seconds < 60 and 7 or (max_seconds < 3600 and 4 or 0))
-		if trim_length > 0 then
-			local has_minus = seconds < 0
-			human = string.sub(human, trim_length + (has_minus and 1 or 0))
-			if has_minus then human = '-' .. human end
-		end
-	end
 	return human
 end
```

**Reply by ElvioG** [reply]
> Here is how to change that
> 


Yes, thank you, I was just looking at this function

**Comment by bananakid** [comment]


**Reply by tomasklaen** [reply]
You'll have to modify the "Window controls" section of the `TopBar:render()` function:

https://github.com/tomasklaen/uosc/blob/7fbcff41c28ac013d0c844053ff0ce90f74a0ed6/src/uosc/elements/TopBar.lua#L234-L271

**Reply by tomasklaen** [reply]
ASS color format is BGR instead of RGB for some reason.
